pkg/peertls/extensions/extensions.go

```diff
@@ -0,0 +1,211 @@
+	// for the current TLS cert chain.
+	RevocationExtID = ExtensionID{2, 999, 1, 2}
+
+	// Error is used when an error occurs while processing an extension.
+	Error = errs.Class("extension error")
+
+	// ErrVerifyCASignedLeaf is used when a signed leaf extension signature wasn't produced
+	// by any CA in the whitelist.
+	ErrVerifyCASignedLeaf = Error.New("leaf not signed by any CA in the whitelist")
+	// ErrUniqueExtensions is used when multiple extensions have the same Id
+	ErrUniqueExtensions = Error.New("extensions are not unique")
+)
+
+// ExtensionID is an alias to an `asn1.ObjectIdentifier`.
+type ExtensionID asn1.ObjectIdentifier
+
+// Config is used to bind cli flags for determining which extensions will
+// be used by the server
+type Config struct {
+	Revocation          bool `default:"true" help:"if true, client leaves may contain the most recent certificate revocation for the current certificate"`
+	WhitelistSignedLeaf bool `default:"false" help:"if true, client leaves must contain a valid \"signed certificate extension\" (NB: verified against certs in the peer ca whitelist; i.e. if true, a whitelist must be provided)"`
+}
+
+// Options holds common options for use in handling extensions.
+type Options struct {
+	PeerCAWhitelist []*x509.Certificate
+	RevDB           RevocationDB
+}
+
+// HandlerFactories is a collection of `HandlerFactory`s for convenience.
+// Defines `Register` and `WithOptions` methods.
+type HandlerFactories []*HandlerFactory
+
+// HandlerFactory holds a factory for a handler function given the passed `Options`.
+// For use in handling extensions with the corresponding ExtensionID.
+type HandlerFactory struct {
+	id      *ExtensionID
+	factory HandlerFactoryFunc
+}
+
+// HandlerFactoryFunc is a factory function used to build `HandlerFunc`s given
+// the passed options.
+type HandlerFactoryFunc func(options *Options) HandlerFunc
+
+// HandlerFunc takes an extension and the remote peer's certificate chains for
+// use in extension handling.
+type HandlerFunc func(pkix.Extension, [][]*x509.Certificate) error
+
+// HandlerFuncMap maps an `ExtensionID` pointer to a `HandlerFunc`.
+// Because an `ExtensionID` is a pointer , one can use a new pointer to the same
+// asn1 object ID constant to store multiple `HandlerFunc`s for the same
+// underlying extension id value.
+type HandlerFuncMap map[*ExtensionID]HandlerFunc
+
+func init() {
+	// NB: register all handlers defined in this file.
+	AllHandlers.Register(
+		CAWhitelistSignedLeafHandler,
+	)
+}
+
+// NewHandler builds a `Handler` pointer from an `ExtensionID` and a `handlerFactoryFunc`.
+func NewHandler(id *ExtensionID, handlerFactory HandlerFactoryFunc) *HandlerFactory {
+	return &HandlerFactory{
+		id:      id,
+		factory: handlerFactory,
+	}
+}
+
+// AddSignedCertExt generates a signed certificate extension for a cert and attaches
+// it to that cert.
+func AddSignedCertExt(key crypto.PrivateKey, cert *x509.Certificate) error {
```

Comment:
`extensions.AddSignedCertExt` is the full name, so maybe something like `extensions.AddSignedCert` would be nicer, because the `Ext` part is already part of the package name.